### PR TITLE
UI: Grenade Throw Arc

### PR DESCRIPTION
**Pull request by EntranceJew**
what's a parabola you, eh?

### CODE REVIEW

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 1]
```diff
@@ -84,6 +85,107 @@ function SWEP:SecondaryAttack()
...
+local function PositionFromPhysicsParams(P, V, G, T)
+	local A = G * physenv.GetGravity()
+
+	return P + (V * T + 0.5 * A * T ^ 2)
+end
+
+if CLIENT then
+	---
+	-- @realm client
+	local trajectory_ui = CreateConVar("ttt2_grenade_trajectory_ui", 0, FCVAR_ARCHIVE)
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:103]: to be in style with our other convar variables, this should probably be `cvEnableTrajectoryUI`

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 2]
```diff
@@ -84,6 +85,107 @@ function SWEP:SecondaryAttack()
...
+		local P = pos - ply:EyePos() + src
+		local V = thr
+		local G = 1
+		local step = 0.005
+		local lastpos = PositionFromPhysicsParams(P, V, G, step)
+		local frac = (SysTime() % 1) * 2
+		local i = frac > 1 and 1 or 0
+		frac = frac - math.floor(frac)
+		local arc_color = appearance.ShouldUseGlobalFocusColor() and appearance.GetFocusColor() or self:GetOwner():GetRoleColor()
+		local alpha = math.Round(GetConVar("ttt_crosshair_opacity"):GetFloat() * 255)
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:127]: could you please give the variables clearer names? I know that in base TTT2 there are also quite a few badly named variables, but I have no idea what P, V, G and i are.

Stuff like thr should probably be thrower? or is it something different than the thrower entity?

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 3]
```diff
@@ -84,6 +85,107 @@ function SWEP:SecondaryAttack()
...
+
+		render.SetColorMaterial()
+		cam.Start3D(EyePos(), EyeAngles())
+
+		for T = step * 2, 1, step do
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:132]: capital T is the time for a full period, is this the indended use here? or what does this variable mean?

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 4]
```diff
@@ -84,6 +85,107 @@ function SWEP:SecondaryAttack()
...
+			pos = PositionFromPhysicsParams(P, V, G, T)
+			local t = util.TraceLine({
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:134]: is the T related to the t?

**Review by TimGoll — CHANGES_REQUESTED** [review]
First: The code is fine and I like the size of this PR, good work. But the variables in this PR are badly named. If you revise the names to be a bit more telling, then it would be really nice. Thank you

**Comment by EntranceJew** [comment]
i dunno

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 5]
```diff
@@ -84,6 +85,123 @@ function SWEP:SecondaryAttack()
...
+		local stepSize = 0.005
+		local gravityMultiplier = 1
+
+		local owner = self:GetOwner()
+		local currentIterationPosition, _ = self:GetViewModelPosition(owner:EyePos(), owner:EyeAngles())
+		local launchPosition, launchVelocity = self:GetThrowVelocity()
+		currentIterationPosition = currentIterationPosition - ply:EyePos() + launchPosition
+		local previousIterationPosition = PositionFromPhysicsParams(currentIterationPosition, launchVelocity, gravityMultiplier, stepSize)
+
+		--
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:123]:
```suggestion
```

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 6]
```diff
@@ -84,6 +85,123 @@ function SWEP:SecondaryAttack()
...
+		local fractionalFrameTimeForDashedLineSegmentLengthPurposes = (SysTime() % 1) * 2
+		local evenOrOddFrameForDashedLineDrawingPurposes = fractionalFrameTimeForDashedLineSegmentLengthPurposes > 1 and 1 or 0
+		fractionalFrameTimeForDashedLineSegmentLengthPurposes = fractionalFrameTimeForDashedLineSegmentLengthPurposes - math.floor(fractionalFrameTimeForDashedLineSegmentLengthPurposes)
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:126]: you know these variable names are ridiculous

**Review by TimGoll — CHANGES_REQUESTED** [review]

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua (hunk 6 above)
Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:125]: even so, i don't know any middleground between what they were and what they are

Comment (TimGoll) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:125]: Dunno, maybe something like 'fractionalFrameTime'? It doesn't have to describe it in detail, it just shouldn't be called `a`

Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua]: ok

Comment (TimGoll) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua]: the line below as well, then it is fine with me

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 7]
```diff
@@ -192,8 +334,8 @@ function SWEP:Throw()
 		tang.p = math.Clamp(tang.p,-90,90) -- Makes the grenade not go backwards :/
 
 		local vel = math.min(800, (90 - tang.p) * 6)
-		local thr = tang:Forward() * vel + ply:GetVelocity()
-		self:CreateGrenade(src, Angle(0,0,0), thr, Vector(600, math.random(-1200, 1200), 0), ply)
+		local force = tang:Forward() * vel + ply:GetVelocity()
+		self:CreateGrenade(src, Angle(0,0,0), force, Vector(600, math.random(-1200, 1200), 0), ply)
```
Comment (saibotk) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:338]: I think you want to replace the code above with the `GetThrowVelocity` function

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 8]
```diff
@@ -84,6 +85,122 @@ function SWEP:SecondaryAttack()
...
+	end
+
+	---
+	-- @param Entity vm
+	-- @param Weapon weapon
+	-- @param Player ply
+	-- @realm client
+	function SWEP:PostDrawViewModel(vm, weapon, ply)
+		if not cvEnableTrajectoryUI:GetBool() then return end
+		self:DrawDefaultThrowPath(ply)
```
Comment (saibotk) [inline gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua:200]:
```suggestion
		if cvEnableTrajectoryUI:GetBool() then
		    self:DrawDefaultThrowPath(ply)
		end
```

**Review by saibotk — CHANGES_REQUESTED** [review]

**Review by TimGoll — APPROVED** [review]
For some reason I rerequested a review by me, well then, it still looks good to me

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua (hunk 7 above)
Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua]: fixed, good spotting

File: gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua [hunk 9]
```diff
@@ -84,6 +85,122 @@ function SWEP:SecondaryAttack()
...
+		local fr = frac ^ 0.5
+		return ColorAlpha(from, Lerp(fr, 0, math.min(max, 255)))
+	end
+
+	---
+	-- @param Player ply
+	-- @realm client
+	function SWEP:DrawDefaultThrowPath(ply)
+		local stepSize = 0.005
+		local gravityMultiplier = 1
```
[the remark this answers is not in the document]
Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbasegrenade.lua]: ok

**Review by saibotk — APPROVED** [review]

**Pull request merged**